From: Vitaliy Filippov <vitali@st-filippov.office.custis.ru>
Subject: [PATCH] t/001-bug53818-empty-heading-crutch
Depends on: Y-000-shift-included-templates-heading-levels.diff

A logical patch which removes empty headings from parser output
instead of taking '==' as their text and all the remaining ='s
as heading marker.

Signed-off-by: Vitaliy Filippov <vitali@st-filippov.office.custis.ru>

---
 includes/parser/Preprocessor_DOM.php |   27 +++++++++++++++------------
 1 files changed, 15 insertions(+), 12 deletions(-)

Index: includes/parser/Preprocessor_DOM.php
===================================================================
--- includes/parser/Preprocessor_DOM.php	(revision 54846)
+++ includes/parser/Preprocessor_DOM.php	(working copy)
@@ -986,20 +1018,23 @@
 
                     # Insert a heading marker only for <h> children of <root>
                     # This is to stop extractSections from going over multiple tree levels
-                    if ( $contextNode->parentNode->nodeName == 'root'
-                      && $this->parser->ot['html'] )
+                    if ($s != str_repeat('=', $contextNode->getAttribute('level')*2))
                     {
-						# Insert heading index marker
-						$headingIndex = $contextNode->getAttribute( 'i' );
-						$titleText = $this->title->getPrefixedDBkey();
-						$this->parser->mHeadings[] = array( $titleText, $headingIndex );
-						$serial = count( $this->parser->mHeadings ) - 1;
-						$marker = "{$this->parser->mUniqPrefix}-h-$serial-" . Parser::MARKER_SUFFIX;
-						$count = $contextNode->getAttribute( 'level' );
-						$s = substr( $s, 0, $count ) . $marker . substr( $s, $count );
-						$this->parser->mStripState->general->setPair( $marker, '' );
+                        if ( $contextNode->parentNode->nodeName == 'root'
+                          && $this->parser->ot['html'] )
+                        {
+							# Insert heading index marker
+							$headingIndex = $contextNode->getAttribute( 'i' );
+							$titleText = $this->title->getPrefixedDBkey();
+							$this->parser->mHeadings[] = array( $titleText, $headingIndex );
+							$serial = count( $this->parser->mHeadings ) - 1;
+							$marker = "{$this->parser->mUniqPrefix}-h-$serial-" . Parser::MARKER_SUFFIX;
+							$count = $contextNode->getAttribute( 'level' );
+							$s = substr( $s, 0, $count ) . $marker . substr( $s, $count );
+							$this->parser->mStripState->general->setPair( $marker, '' );
+						}
+						$out .= $s;
 					}
-					$out .= $s;
 				} else {
 					# Generic recursive expansion
 					$newIterator = $contextNode->childNodes;
